sycl/doc/SharedLibraries.md

```diff
@@ -233,6 +233,45 @@ This native device image is then added to the cache to avoid symbol resolution,
 compilation, and linking for any future attempts to invoke kernels from this
 device image.
 
+#### DPC++ runtime plugin interface (PI) changes
+
+Before creating a program the function `piextDeviceSelectBinary` is used to
+choose the most appropriate device image. It is possible that not all backends
+have possibility to link device images of particular format at run-time. So,
+in presence of dynamic linking the `piextDeviceSelectBinary` function should be
+extended, so it chooses the appropriate device image
```

Comment:
> One reason is that most of the logic will be the same for all plugins - finding transitive closure of modules to be linked, with some may having multiple formats, then removing formats for which there is no linking support.

Please note that `piextDeviceSelectBinary` is only intended to be used to find the **first** device image among several which contain a kernel which is launched by user. Basically here we decide on the format of all device images which will be used in online linking (if necessary): once the first image is known, we lookup its dependencies in RT, but only looking into device images of the same type.

Reply:
Why do not allow linking of binaries of different types, SPIR-V linked with native code?

Reply:
After offline discussion it become clear that we don't have problems with SPIR-V linking support. We may have some problems with "native" code linking, but I realized that we don't need described change in `piextDeviceSelectBinary` . Right now AOT compilers don't allow to compile device code with unresolved symbols.
If AOT compiler for particular backend allowed to generate some code with unresolved symbols, then this backend should be able to link this code. In other words it is not possible to meet device image with "native code" format without ability to link it.
It also seems that we could allow linking of binaries of different types, however we need to define mechanism for case when both formats are available. So, when both formats for dependency resolution are available I suggest the following:
- If target image is SPIR-V, then runtime should pick SPIR-V device images to resolve dependencies, so ability for link time optimizations won't be lost
- If target image is native code, then runtime should pick native code device images to resolve dependencies, so additional JIT is not required.

@kbobrovs , @AlexeySachkov , @smaslov-intel do you agree? If yes, I'll update the doc accordingly, otherwise please share your thoughts.

Reply:
> Please note that piextDeviceSelectBinary is only intended to be used to find the first device image among several which contain a kernel which is launched by user. Basically here we decide on the format of all device images which will be used in online linking (if necessary): once the first image is known, we lookup its dependencies in RT, but only looking into device images of the same type.

With dynamic runtime linkage this logic should be extended. Your suggestion:
- piextDeviceSelectBinary selects single preferable binary image `A` out of, say [`native_arch_xv1`, `native_arch_xv2`, `SPIRV`], and say, `native_arch_xv2` is selected
- SYCL RT then resolves `A`'s dependencies looking into import lists of only `native_arch_xv2` device binary images, by building a transitive closure of `depends` module relation
If at some step there is no binary with needed `native_arch_xv2` format, linkage fails.

My suggestion is to repeat that process for all available formats in the order of preference, until linkage succeeds. The order of preference can be determined by `piextDeviceSelectBinary` - in the above example, `native_arch_xv2` should be removed after linkage failure, and at step 2 only [`native_arch_xv1`, `SPIRV`] is on the menu for `piextDeviceSelectBinary `. `native_arch_xv1` may fail too, and only `SPIRV` may succeed.
Plugin does not have enough scope to decide that SPIRV should eventually be selected, this is SYCL RT responsibility.

Does that sound reasonable?

Reply:
> Right now AOT compilers don't allow to compile device code with unresolved symbols.

I don't think this is true for GPU native code. zeDynamicLink wouldn't have existed otherwise. Looking forward, other AOT compilers for targets should allow unresolved symbols. So the design should accommodate that.

> It also seems that we could allow linking of binaries of different types, however we need to define mechanism for case when both formats are available.

Are you taking about mixing different types in single module's linkage? That's is a really good point, which I also thought about. Say, we have module A in `native_arch_xv1` format and B in `SPIRV` format, and A has unresolved symbols defined in B. Then we may resolve it by first compiling B to  `native_arch_xv1` if possible, then doing linkage. But I would just mention this as a TODO and leave for future incremental improvement.

Regarding the format selection logic - I described my view above. Hope this makes sense.

Reply:
> I don't think this is true for GPU native code. zeDynamicLink wouldn't have existed otherwise

I believe it is true and there is an open bug report.  My understanding is that we have never used `zeDynamicLink` in DPC++ yet because `zeModuleCreate` always fails when there is an unresolved symbol.  Obviously, this is a critical bug for this new proposed feature if we want to support AOT mode, so we should call out the dependency and make sure there is a commitment to get it fixed.

> Regarding the format selection logic - I described my view above.

I think you only address the case where all the formats are variants of different **native** code images.  What if there is both a SPIR-V image and a native code image?  Before this shared library proposal, the choice was easy.  It was always better to choose the native code image in such a case (because it eliminates the JIT overhead).  Now, there is a trade-off, so the choice isn't clear:

* You could select all the native code images and link them with `zeDynamicLink`.  This avoids JIT overhead, but loses out on inter-module optimizations.
* You could  select all the SPIR-V images and compile/link them with a proposed extension to `zeModuleCreate`.  This suffers a JIT overhead, but potentially results in faster device code due to inter-module optimizations.

How do you decide which one the user wants?

A reasonable option might be to look at the image containing the kernel function that we want to launch.  If that image has a native code version of the kernel function, assume the user prefers AOT mode and link with `zeDynamicLink`.  If the kernel function only exists in SPIR-V, prefer SPIR-V for all its dependencies too and compile everything with `zeCreateModule`.

Reply:
> I believe it is true and there is an open bug report.

OK, thanks for reminding. This won't be true soon. Linkage of Gen native code is something invoke_simd will (at least initially) rely on, so this bug should be fixed for sure.

> I think you only address the case where all the formats are variants of different native code images. 

No, I considered the case "piextDeviceSelectBinary selects single preferable binary image A out of, say [native_arch_xv1, native_arch_xv2, SPIRV], and say, native_arch_xv2 is selected". piextDeviceSelectBinary decides about format preferences, SYCL RT sees which format is possible at all. It should be possible to override behavior of piextDeviceSelectBinary. Either via env var or via an extension API from the program.

> A reasonable option might be to look at the image containing the kernel function that we want to launch. 

There can be multiple binary images of different formats containing definition of the same kernel. I described above the process which would allow to make good decision, IMO.

Even in case of single available format for a kernel - native - there are options. Native could be linked with SPIRV after the SPIRV is compiled to native.

> If the kernel function only exists in SPIR-V, prefer SPIR-V for all its dependencies too and compile everything with zeCreateModule.

In the existing design the decision about format preferences is `piextDeviceSelectBinary` responsibility . Each plugin can decide as it sees fit. Maybe taking into account env var or requests via extension API calls I mentioned above.

Reply:
> No, I considered the case "piextDeviceSelectBinary selects single preferable binary image A out of, say [native_arch_xv1, native_arch_xv2, SPIRV], and say, native_arch_xv2 is selected". piextDeviceSelectBinary decides about format preferences

Oh, sorry.  I missed that SPIR-V was in that list.  I assume that "binary image A" here is the image that contains the kernel function that the application wants to launch?

I think we plan for the device library fall-back code to be distributed in SPIR-V format.  Once this is the case, many kernels may require linking with this SPIR-V module.  Once we support shared libraries with AOT (native) modules, it will be common to link these AOT modules with the SPIR-V device library fall-back code.  Therefore, we will need to support mixed native / SPIR-V linking anyways.

Reply:
> I assume that "binary image A" here is the image that contains the kernel function that the application wants to launch?

yes

> Therefore, we will need to support mixed native / SPIR-V linking anyways.

I agree with that. I guess the main question  is about timing - do you suggest to implement this in the first step? My suggestion was:

> But I would just mention this as a TODO and leave for future incremental improvement.

Reply:
I think the situation when mixed linking is required can be pretty common. For example because compilation of SPIR-V is not enforced when user requested AOT. However I don't have a strong opinion regarding timings.
The algorithm described here by @kbobrovs :
> piextDeviceSelectBinary selects single preferable binary image A out of, say [native_arch_xv1, native_arch_xv2, SPIRV], and say, native_arch_xv2 is selected
SYCL RT then resolves A's dependencies looking into import lists of only native_arch_xv2 device binary images, by building a transitive closure of depends module relation
If at some step there is no binary with needed native_arch_xv2 format, linkage fails.
My suggestion is to repeat that process for all available formats in the order of preference, until linkage succeeds. The order of preference can be determined by piextDeviceSelectBinary - in the above example, native_arch_xv2 should be removed after linkage failure, and at step 2 only [native_arch_xv1, SPIRV] is on the menu for piextDeviceSelectBinary . native_arch_xv1 may fail too, and only SPIRV may succeed.

may need changes in case we want to use mixed linking. With mixed linking, linkage of device image `A` of `native_arch_xv2`  format shouldn't fail if there is SPIR-V image that contains required export.

If we are not going to target mixed linking in the first iteration, then I'll update the doc with this algorithm without changes, since it seems reasonable for me.
@kbobrovs please share your opinion.

Reply:
Yes, the searching algorithm itself is more or less clear to me.
However I've got one more question.
> if it can link native format

I guess there is a way (at least in OpenCL) to query whether compiler and linker available. However I'm not sure there is any option to ask if it can link native binaries. My original idea was that PI plugin implementation for concrete backend will answer this question, however If I understand your previous comments correctly, it shouldn't be there. So, how we can understand in SYCL RT whether target backend can or cannot link native binaries?

Reply:
> My original idea was that PI plugin implementation for concrete backend will answer this question, however If I understand your previous comments correctly, it shouldn't be there.

Not sure which my comment makes you think so. I think this is exactly the way it should be implemented. Some new PI function to enumerate BE capabilities. It should be fine-grained enough to distinguish between devices covered by a plugin, between different formats.
@smaslov-intel, @romanovvlad - do you have ideas here?

Note that SYCL spec already requires RT to be able to answer device aspect queries, such as `online_linker`, which adds similar requirements.

Reply:
Maybe we can extend `pi_device_info` enum (https://github.com/intel/llvm/blob/07b279653d9e0808f21fe7dc11b6efba7887686c/sycl/include/CL/sycl/detail/pi.h#L289) with a new value that will answer the question regarding possibility of native binaries linkage. Something like `PI_DEVICE_INFO_DYNAMIC_LINKER_AVAILABLE`.
So we can re-use `piDeviceGetInfo` API.

Reply:
Will it cover:
>  It should be fine-grained enough to distinguish between devices covered by a plugin, between different formats.

?

Reply:
An example: BE can link SPIRV, but can't link native. Or vice versa. If we can make `piDeviceGetInfo` cover this - I'm OK. Maybe via two new kinds of info instead of one.

Reply:
If we do add a query interface then extending existing `piDeviceGetInfo` does sound like the right approach. Let me ask however, what would SYCL RT do differently if backend reports no support for dynamic linking? Can this something be based of `piProgramLink` returning an error?

Reply:
Good question.
The scenario can be like this: SYCL RT sees two binary image formats available for a function. It also sees that both have dynamic dependencies and need to be linked, Using `piProgramLink` to detect whether any of the two formats can be linked may bring significant overhead - first we need to prepare data, then  `piProgramLink` may take some time to report an error. So it seems much cleaner way not to try to use capabilities not supported by the BE.

Reply:
> what would SYCL RT do differently if backend reports no support for dynamic linking?

It will affect choice of device images. For example: there is a request from user to run particular kernel. There is two device images available that contain required kernel name: in SPIR-V format and native. However kernel and therefore device images that contain this kernel have import symbol, i.e. dependency. The required symbol is defined by some other device image, say it exists only in SPIR-V format. If backend don't support dynamic linking of native binaries, then RT will use SPIR-V image that contains kernel and link it on SPIR-V level with device image with required symbol. Otherwise it could use native format to minimize JIT cost. Or another example, if device image with required symbol exists only in native format and device reports no support for native binaries linking, then RT will emit an error, since there is no way to resolve dependencies and therefore get an executable program.
> BE can link SPIRV, but can't link native. Or vice versa. If we can make piDeviceGetInfo cover this - I'm OK.

Inside PI plugin we can understand on which device we are working on, right? I guess if yes, then we can just make some switch by device name/type and based on current information about support return positive or negative answer. Yes, seems a bit hardcoded, but doesn't require OpenCL extensions.

Reply:
> Inside PI plugin we can understand on which device we are working on, right? I guess if yes, then we can just make some switch by device name/type and based on current information about native linking support and return positive or negative answer. Yes, seems a bit hardcoded, but doesn't require OpenCL extensions.

Hm, I don't see what you mean. Only RT knows what formats are available for particular device and what type of linkage (SPIRV or native) might be needed. If BE supports linking of just one of the two (or three) formats, how it can report which one is supported via single boolean property? We may make `piDeviceGetInfo` with `PI_DEVICE_INFO_DYNAMIC_LINKER_AVAILABLE` argument return a list of bools, one per linkage type - that would work too.

Reply:
BTW, is this backend specific info or really device specific capability? If it the earlier (very likely)  then maybe the proper way would be to add new "extension" into `piPlatformGetInfo/PI_PLATFORM_INFO_EXTENSIONS` instead.

Reply:
> is this backend specific info or really device specific capability?

I would say that it is backend capability, but doesn't each device have specific backend? I mean, we have OpenCL CPU backend and OpenCL GPU backend.
I'm not sure that using of platform info is a proper way. If I understand it correct, in general in OpenCL platform may contain several devices. And we can get a situation when platform contains several devices, but support for native code linking is implemented not for all of them. I guess for the same reason OpenCL defines queries like "compiler/linker_available" for particular device, but not platform.

> I don't see what you mean. 

My idea was that `PI_DEVICE_INFO_DYNAMIC_LINKER_AVAILABLE` will answer only the question "if native code linking supported?". However it is not flexible if we for example may have several "close to native" but linkable device code formats.

> We may make piDeviceGetInfo with PI_DEVICE_INFO_DYNAMIC_LINKER_AVAILABLE argument return a list of bools, one per linkage type - that would work too.

I like this idea. We can even return a list of supported formats as a string with separators, since return parameter of `piDeviceGetInfo` has `void*` type.
```
__SYCL_EXPORT pi_result piDeviceGetInfo(pi_device device,             
                                        pi_device_info param_name,    
                                        size_t param_value_size,      
                                        void *param_value,            
                                        size_t *param_value_size_ret);
```
As I can see similar is used in OpenCL, example `CL_PLATFORM_EXTENSIONS`. Something like
| pi_device_info | Description |
| --- | --- |
| `PI_DEVICE_INFO_LINKING_FORMATS` | Return type: char[]. Returns space separated list of devic images formats that can be linked at runtime by backend implementation for this device |

Reply:
Another alternative is to ask the BE/device if it supports particular [format](https://github.com/intel/llvm/blob/47aeafa8a180ee2f5dfb741f54defbe25e8a78a7/sycl/include/CL/sycl/detail/pi.h#L660) RT found in the binary image descriptor. 

It seems more convenient to use, as SYCL RT does not need to iterate through the list of returned formats, no need to manage storage for the list.

Reply:
> If I understand it correct, in general in OpenCL platform may contain several devices.

This is good point. Per-device granularity is more accurate.

Reply:
> Another alternative is to ask the BE/device if it supports particular format RT found in the binary image descriptor.

I'm not sure that I see how it can be done through existing `piDeviceGetInfo` API, so it will probably require some new PI function. @smaslov-intel , do you have any preference?

Reply:
Me preference would be not to add new PI interface for this. We should extend `piDeviceGetInfo`. Note, that there is existing `PI_DEVICE_INFO_EXTENSIONS` to query device extensions. So, maybe, just define few new extensions, and check for them in SYCL RT. They should be documented in pi.h then.

Reply:
Ok, I've made a change to PI section, please review. If this is ok, I can prepare a PR to pi.h.